### ISSUE TITLE
Support for synchronised crawls

### ISSUE THREAD

**Issue by bkrumnow** [mozilla/OpenWPM#766]
It looks like support for synchronised crawls was removed in 560c6579a127d1a7d8b47fc5204d93e26cfa2a08 and e47e0eef31170d9b733a652ada1a5c4d3ff8a742.
The example in the [docs](https://github.com/mozilla/OpenWPM/wiki/Platform-Demo#running-a-simple-crawl) proposes using `index='**'` to synchronise crawls, which appears to be not working anymore.

So, I am wondering if it is still supported and how to set it up?

Thanks for any hints that point me in the right direction.

**Comment by vringar** [comment]
I'm sorry for our wiki being terribly out of date we are currently trying to move this into the repository with #716 
Can you tell me your use case, so we can work out a setup that works for what you want to do?

**Comment by bkrumnow** [comment]
Thanks for taking this up. My use case is: Run multiple crawlers that visit the same website at the very same moment.

**Comment by vringar** [comment]
Okay, then I'd suggest you modify `demo.py` by first setting NUM_BROWSERS to however many browsers you want to have an replace
https://github.com/mozilla/OpenWPM/blob/6a8bd88b87f3dea93f84ac07017e045859d989e3/demo.py#L42-L56
with something like

```python
site = "https://mysite.example.com"

for i in range(NUM_BROWSERS):
    # Parallelize sites over all number of browsers set above.
    command_sequence = CommandSequence.CommandSequence(
        site,
        reset=True,
        callback=lambda success, val=i: print("CommandSequence {} done".format(val)),
    )

    # Start by visiting the page
    command_sequence.get(sleep=3, timeout=60)

    # Run commands across the three browsers (simple parallelization)
    manager.execute_command_sequence(command_sequence, index=i)
```
If this is simultaneous enough for you, we can leave it at that, but we still have a lot of options to explore afterwards.

**Comment by vringar** [comment]
(This is basically what synchronized crawls looked like under the hood)

**Comment by bkrumnow** [comment]
Works perfect! Thanks :) 👍

**Comment by bkrumnow** [comment]
Hey,
Apologies on re-opening this, but I wanted to dive a bit deeper. I saw that crawlers start at a time, but loose synchronicity over time. 

> but we still have a lot of options to explore afterwards.

Is there an option where crawlers start the get command for the same site at the very same time?

Thanks a lot in advanced! :)

**Comment by vringar** [comment]
I'll write up a tutorial and some sample code, to show how I think this could be done.
My basic idea is:
Insert a custom command that utilizes  [a barrier](https://docs.python.org/3/library/threading.html#barrier-objects) to sync all browsers and then run the GetCommand afterwards.

**Comment by vringar** [comment]
Should I expand on this idea further or do you want to try and do it on your own.
I could either write this up as a tutorial or we could set up a meeting and develop this together.

**Comment by bkrumnow** [comment]
Thanks for the fast reply.
Both works for me. I will direct message you to discuss it.

**Comment by bkrumnow** [comment]
Very first insights from testing your proposed solution (I used OpenWPM 0.13.0): It works well! :)

However, it can lead to deadlocks (as any other solution potentially can). I run into deadlocks when using timeouts for the GetCommand that were larger than for the lockstep command. In my case, some browsers run out of time waiting for other browsers while the other browsers were still loading the previously scanned site.

It is possible to avoid this either by warning users or by calculating the timeout based on the other commands in the command sequence. 

I hope that helps. I will run more tests over the next 3-4 days and get back to you if there is more to report.

**Comment by bkrumnow** [comment]
Hey,


I have experimented a bit more. I observed that browsers sometimes crash during startup and break the barrier. The barrier does not recover on its own and needs resetting to prevent deadlocks. 


Something like this can do the job (v0.13.0 style):

```python
from multiprocessing import Barrier

BARRIER = Barrier(8)

def lock_step(barrier_timeout, **kwargs):
    """This commands aims to sync up a group of
    browsers that are all executing this command"""
    try:
        BARRIER.wait(barrier_timeout)
    except:
        BARRIER.reset()
        print("A barrier reset was triggered")
        raise
```


Still, it does not prevent all problems. In one run, I saw that all of my browsers were scanning the same site, except for one browser that was one site ahead. Another problem is that when the barrier breaks, all browsers will stop, and the data for this site is lost. Still, this is rare; in my case, this happened in 0,5% of all crawls.


So overall, this solution can do the job but needs more polishing in case recovery is necessary.

From my site, we can close the ticket. Thanks again for your awesome help!
